### PR TITLE
e2e: Use libvmi and libnet directly instead NewRandomFedoraVMI

### PR DESCRIPTION
**Pull request by oshoval**
### What this PR does
Use libvmi and libnet directly instead `NewRandomFedoraVMI`

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by oshoval** [comment]
/hold

until https://github.com/kubevirt/kubevirt/pull/11217 is merged

/test pull-kubevirt-e2e-k8s-1.29-sig-network

**Comment by oshoval** [comment]
/hold cancel

rebased

**Comment by oshoval** [comment]
fixed `pull-kubevirt-check-unassigned-tests` failure

File: tests/migration/migration.go [hunk 1]
```diff
@@ -3593,3 +3590,7 @@ func runStressTest(vmi *v1.VirtualMachineInstance, vmsize string, stressTimeoutS
 		time.Sleep(15 * time.Second)
 	}
 }
+
+func newVirtualMachineInstanceWithFedoraContainerDisk() *v1.VirtualMachineInstance {
```
Comment (EdDev) [inline tests/migration/migration.go:3594]: It is prefered to do this in a separate commit in order to reason for it.

Is there an explanation why this is not inlined now?

File: tests/monitoring/vm_monitoring.go [hunk 2]
```diff
@@ -25,6 +25,7 @@ import (
 	"time"
 
 	"kubevirt.io/kubevirt/tests/libmigration"
+	"kubevirt.io/kubevirt/tests/libnet"
```
Comment (EdDev) [inline tests/monitoring/vm_monitoring.go:28]: Please place this at the last block, it seems to fit there.

File: tests/portforward_test.go [hunk 3]
```diff
@@ -26,6 +26,8 @@ import (
 	"time"
 
 	"kubevirt.io/kubevirt/tests/decorators"
+	"kubevirt.io/kubevirt/tests/libnet"
+	"kubevirt.io/kubevirt/tests/libvmi"
```
Comment (EdDev) [inline tests/portforward_test.go:30]: Seems to fit in the last block where all the other `tests` are located.

File: tests/storage/restore.go [hunk 4]
```diff
@@ -7,6 +7,7 @@ import (
 	"time"
 
 	"kubevirt.io/kubevirt/tests/framework/checks"
+	"kubevirt.io/kubevirt/tests/libnet"
```
Comment (EdDev) [inline tests/storage/restore.go:10]: Please locate it at the last block where all the other tests libs are now.

File: tests/storage/snapshot.go [hunk 5]
```diff
@@ -7,6 +7,7 @@ import (
 	"time"
 
 	"kubevirt.io/kubevirt/tests/decorators"
+	"kubevirt.io/kubevirt/tests/libnet"
```
Comment (EdDev) [inline tests/storage/snapshot.go:10]: Please move this to the last block.

File: tests/subresource_api_test.go [hunk 6]
```diff
@@ -25,6 +25,7 @@ import (
 	"time"
 
 	"kubevirt.io/kubevirt/tests/decorators"
+	"kubevirt.io/kubevirt/tests/libnet"
```
Comment (EdDev) [inline tests/subresource_api_test.go:28]: Please move this to the last block.

File: tests/swap_test.go [hunk 7]
```diff
@@ -27,6 +27,8 @@ import (
 	"time"
 
 	"kubevirt.io/kubevirt/tests/libmigration"
+	"kubevirt.io/kubevirt/tests/libnet"
+	"kubevirt.io/kubevirt/tests/libvmi"
```
Comment (EdDev) [inline tests/swap_test.go:31]: Fits in the 55-57 block.

File: tests/vm_state_test.go [hunk 8]
```diff
@@ -6,6 +6,7 @@ import (
 	"time"
 
 	"kubevirt.io/kubevirt/tests/libmigration"
+	"kubevirt.io/kubevirt/tests/libnet"
```
Comment (EdDev) [inline tests/vm_state_test.go:9]: Please place it in the last block where other libs exist.

File: tests/vmi_configuration_test.go [hunk 9]
```diff
@@ -33,6 +33,7 @@ import (
 	"kubevirt.io/kubevirt/pkg/pointer"
 
 	"kubevirt.io/kubevirt/tests/decorators"
+	"kubevirt.io/kubevirt/tests/libnet"
```
Comment (EdDev) [inline tests/vmi_configuration_test.go:36]: Please place it in the last block where the other libs exist.

File: tests/vmi_headless_test.go [hunk 10]
```diff
@@ -28,6 +28,8 @@ import (
 
 	"kubevirt.io/kubevirt/tests/framework/kubevirt"
 	"kubevirt.io/kubevirt/tests/framework/matcher"
+	"kubevirt.io/kubevirt/tests/libnet"
+	"kubevirt.io/kubevirt/tests/libvmi"
```
Comment (EdDev) [inline tests/vmi_headless_test.go:32]: Please move it to the last block.

File: tests/vmi_hostdev_test.go [hunk 11]
```diff
@@ -6,6 +6,8 @@ import (
 	"strings"
 
 	"kubevirt.io/kubevirt/tests/decorators"
+	"kubevirt.io/kubevirt/tests/libnet"
+	"kubevirt.io/kubevirt/tests/libvmi"
```
Comment (EdDev) [inline tests/vmi_hostdev_test.go:10]: Same

File: tests/vmi_multiqueue_test.go [hunk 12]
```diff
@@ -25,6 +25,7 @@ import (
 	"fmt"
 
 	"kubevirt.io/kubevirt/tests/decorators"
+	"kubevirt.io/kubevirt/tests/libnet"
```
Comment (EdDev) [inline tests/vmi_multiqueue_test.go:28]: same

File: tests/vmi_vsock_test.go [hunk 13]
```diff
@@ -26,6 +26,7 @@ import (
 	"time"
 
 	"kubevirt.io/kubevirt/tests/libmigration"
+	"kubevirt.io/kubevirt/tests/libnet"
```
Comment (EdDev) [inline tests/vmi_vsock_test.go:29]: same

**Review by EdDev — COMMENTED** [review]
Thank you!

The inline comments are minor, but it should be a small effort to adjust.

File: tests/migration/migration.go (hunk 1 above)
Comment (oshoval) [reply tests/migration/migration.go:3594]: If we try to change it from function pointer to `vmi *v1.VirtualMachineInstance`
and invoke the functions that are there,
i.e `	Entry("[sig-storage][test_id:2226] with ContainerDisk", libvmi.NewFedora(libnet.WithMasqueradeNetworking()...), false),`

I get the following error, since it is part of the `Entry`
```
Assertion or Panic detected during tree construction
tests/migration/migration.go:2161
  Ginkgo detected a panic while constructing the spec tree.
  You may be trying to make an assertion in the body of a container node
  (i.e. Describe, Context, or When).

  Please ensure all assertions are inside leaf nodes such as BeforeEach,
  It, etc.

...

  Learn more at: http://onsi.github.io/ginkgo/#no-assertions-in-container-nodes
```

Comment (EdDev) [reply tests/migration/migration.go:3594]: I do not know what you tried, but you should be able to inline this easily.
`func () *v1.VirtualMachineInstance  {return libvmi.NewFedora(libnet.WithMasqueradeNetworking()...)}`

Comment (oshoval) [reply tests/migration/migration.go:3594]: tried to invoke it
ok to inline like that, sure

Comment (oshoval) [reply tests/migration/migration.go:3594]: tried to invoke it
ok to inline like that, sure it is possible
but it happens twice, so to use it twice ?
i can also keep it lambda ...

**Comment by oshoval** [comment]
1st push - Addressed comments
2nd push - rebased

File: tests/mdev_configuration_allocation_test.go [hunk 14]
```diff
@@ -9,6 +9,8 @@ import (
 	"time"
 
 	"kubevirt.io/kubevirt/tests/decorators"
+	"kubevirt.io/kubevirt/tests/libnet"
```
Comment (oshoval) [inline tests/mdev_configuration_allocation_test.go:12]: to move those as well to last block ? 
maybe even the first one, so it stop adding it there (visual studio code auto added it here)

**Comment by oshoval** [comment]
addressed the first paragraph of https://github.com/kubevirt/kubevirt/pull/11261/files#r1494671044

**Comment by oshoval** [comment]
15:09:40: Error response from daemon: cannot listen on the TCP port: listen tcp4 :38607: bind: address already in use

Hi @brianmcarey 
maybe worth that we consider fixing this on kubevirtci ?
it hits few times a week randomly

**Comment by oshoval** [comment]
rebased

**Review by EdDev — APPROVED** [review]
Thank you!

**Comment by AlonaKaplan** [comment]
/approve

We should change the test config so masquerade will be the default binding. It looks bad that we pass masquerade explicitly to tests that has nothing to do with network.

**Comment by EdDev** [comment]
> We should change the test config so masquerade will be the default binding. It looks bad that we pass masquerade explicitly to tests that has nothing to do with network.

Kubevirt default is "bridge binding", the tests express it.
I am not really sure if all these tests really need masquerade, it may have just been used unintentionally because that helper hided it.
Now it is visible, so it may be relevant to check if it can be dropped.

**Comment by oshoval** [comment]
seems cluster-down failed but tests passed
first time i see it, so no biggie for now
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/11261/pull-kubevirt-e2e-k8s-1.27-sig-compute/1762077379578040320
```
15:58:55: selecting podman as container runtime
15:58:56: Error response from daemon: getting graph driver info "1672b222bee6635410bbe933e864a23bb656d8389cfcf9c8304bc897f42e3118": readlink /var/lib/containers/storage/overlay/l/YMYUV3U6Y7ZO3DIG3HOFQ3HXDY: no such file or directory
make: *** [Makefile:155: cluster-down] Error 1
+ EXIT_VALUE=2
```
